src/game/logic/object/object.cpp

```diff
@@ -2655,51 +2984,468 @@ void Object::Enter_Group(AIGroup *group)
+                m_shroudSighting->m_where = pos;
+                m_shroudSighting->m_playerIndex = index;
+                m_shroudSighting->m_radius = Get_Shroud_Range();
+            }
+        }
+    } else {
+        captainslog_dbgassert(false, "An Object is shrouding, but hasn't unshrouded the last one.");
+    }
+}
+
+void Object::Unshroud()
+{
+    if (!m_shroudSighting->Is_Invalid()) {
+        g_thePartitionManager->Undo_Shroud_Cover(m_shroudSighting->m_where.x,
+            m_shroudSighting->m_where.y,
+            m_shroudSighting->m_radius,
+            m_shroudSighting->m_playerIndex);
+        m_shroudSighting->Reset();
+    }
+}
+
+float Object::Get_Shroud_Range() const
+{
+#ifdef GAME_DEBUG_STRUCTS
+    if (g_theWriteableGlobalData->m_debugVisibility) {
+        Vector3 v(m_shroudRange, 0.0f, 0.0f);
+
+        for (int i = 0; i < g_theWriteableGlobalData->m_debugVisibilityTileCount; i++) {
+            float f1 = i * 1.0f / g_theWriteableGlobalData->m_debugVisibilityTileCount;
+            float angle = (f1 + f1) * GAMEMATH_PI;
+            v.Rotate_Z(angle);
+            Coord3D pos;
+            pos.x = v.X + Get_Position()->x;
+            pos.y = v.Y + Get_Position()->y;
+            pos.z = v.Z + Get_Position()->z;
+            Add_Icon(&pos,
+                g_theWriteableGlobalData->m_debugVisibilityTileWidth,
+                g_theWriteableGlobalData->m_debugVisibilityTileDuration,
+                g_theWriteableGlobalData->m_debugVisibilityTileTargettableColor);
+        }
+    }
+#endif
+    return m_shroudRange;
+}
+
+CountermeasuresBehaviorInterface *Object::Get_Countermeasures_Behavior_Interface()
+{
+    for (BehaviorModule **module = m_allModules; *module != nullptr; module++) {
+        CountermeasuresBehaviorInterface *behavior = (*module)->Get_Countermeasures_Behavior_Interface();
+
+        if (behavior != nullptr) {
+            return behavior;
+        }
+    }
+
+    return nullptr;
+}
+
+const CountermeasuresBehaviorInterface *Object::Get_Countermeasures_Behavior_Interface() const
+{
+    for (BehaviorModule **module = m_allModules; *module != nullptr; module++) {
+        const CountermeasuresBehaviorInterface *behavior = (*module)->Get_Countermeasures_Behavior_Interface();
+
+        if (behavior != nullptr) {
+            return behavior;
+        }
+    }
+
+    return nullptr;
+}
+
+void Object::Score_The_Kill(const Object *victim)
+{
+    Player *victim_player = victim->Get_Controlling_Player();
+
+    if (victim_player->Is_Playable_Side() && !victim->Is_KindOf(KINDOF_IGNORED_IN_GUI)) {
+        Player *player = Get_Controlling_Player();
+
+        if (victim_player != nullptr) {
+            victim_player->Get_Score_Keeper()->Add_Object_Lost(victim);
+        }
+
+        if (this->Get_Relationship(victim) == ENEMIES && player != victim_player) {
+            if (player != nullptr) {
+                player->Get_Score_Keeper()->Add_Object_Destroyed(victim);
```

Comment:
Should this not also be guarded by `Get_Status(OBJECT_STATUS_UNDER_CONSTRUCTION)` ?

Reply:
No.